Session paths handling

csoeller opened the pull request:
Addresses paths handling for saved sessions and adds a GUI interface for `load_extra_datasources`.

**Is this a bugfix or an enhancement?**

Enhancement for general session paths handling and (smallish) bug fix to resolve all paths prior to session saving.

**Proposed changes:**

1. Resolve all file paths when requesting session info via `_get_session_datasources`.
2. Add a GUI interface to load extra datasources.
2. Add the ability make session paths (and some paths in recipes) relative to the `.pvs` session file.

**Details:**

1. Originally file paths were returned as stored with the data source. When started from the command line these may be relative to the shell's current working directory (CWD) and this would then fail, for example, when opening the generated `.pvs` file from the Finder window - because `PYMEVis` will generally start with a quite different CWD. Therefore, we now return fully resolved paths when building the session dictionary. But see also 3.
2. In practice it can be quite a bit more convenient to add relevant extra datasources via the GUI (rather than from the command line), tweak the recipe and then save the whole session in a `.pvs` file. Thus this PR also adds a simple GUI interface to the `pipeline` method `load_extra_datasources`.
3. To optionally use relative file paths when writing session files is likely the proposed addition that is the most experimental in this PR. The motivation arises from two scenarios, the first being our prime motivator. (a) Our files and thus also session files often sit in shared directories, for example in OneDrive or Dropbox. As a result absolute paths differ between machines but relative paths stay the same. (b) We sometimes  move files en-bloc with their session files into a subdirectory, different directory etc. This generally preserves relative paths but clearly not absolute paths.

**about 3 - relative filepaths:**

This PR implements the functionality to write relative paths in sessions (and related recipe modules) but enables this only if explicitly requested by setting the `PYME.config` variable `VisGUI-session_paths_relative` to `true` (it is false by default).

The paths translation is careful in the sense that if paths are already in the state requested (relative or absolute) they are left untouched and if errors are encountered when paths are attempted to be rewritten relative then a warning is issued and the original session dict is returned with paths untouched.

We also enable manual addition of modules with their relevant fields to translate absolute paths in `traits.File` fields in recipes. There may be a better (automagic) way, but in practice we have one or two modules that need this.

Perhaps not everybody will want/like this relative paths functionality but at least in our tests everything seems quite robust and in practice makes life a whole lot easier. In addition, as it is not happening by default we can gain some experience with this feature and see if it is worthwhile.

csoeller committed:
enable path rewriting for sessions upon saving/load

make paths relative to session file
also allow registering modules for which to rewrite recipes

csoeller committed:
GUI method to load extra datasource; relpaths via config

pipeline now guarantees absolute paths for session datasources sessionpaths module tweaks and comments relative path writing in sessions now via config variable VisGUI-session_paths_relative (False by default)

csoeller committed:
GUI method to load extra datasource; relpaths via config

- pipeline now guarantees absolute paths for session
- datasources sessionpaths module tweaks and comments
- relative path writing in sessions now via config variable VisGUI-session_paths_relative (False by default)

csoeller committed:
Merge branch 'session-paths-rewrite' of https://github.com/csoeller/python-microscopy into session-paths-rewrite

csoeller committed:
add some error handling when making relativ paths

csoeller commented on PYME/LMVis/visCore.py:
```diff
@@ -421,6 +422,26 @@ def OnViewInterpTriangles(self,event):
+            else:
+                return result
+        
+        filename = wx.FileSelector("Choose a file to open", 
```
here we should probably add a wildcard flag with all reasonable extensions for which we can currently load datasources (in other words quite a few)...

David-Baddeley commented:
A few potential issue to consider here:

- we already do some relative path handling in the load logic (e.g. `PYME.IO.unifiedIO.local_or_temp_filename` and `PYME.IO.nameUtils.getFullExistingFilename`). I'd be a bit reluctant to add a further, independent, layer of indirection - both due the the added complexity and the fact that it might not play nicely with the existing path handling logic.
- As it stands, the proposed changes will not play nicely with cluster URIs or with the existing handling of paths relative to `PYMEDATADIR`.
- Re-writing the file input to selected recipe modules is super gross - especially requiring recipe modules to register in a seperate place for this to happen. We should really fix file inputs properly and uniformly rather than hack around this in sessions. In general file inputs are an aberration which limits portability of recipes, but it's hard to see an alternative for calibration data that might be used across several recipes. The right answer is probably something along the lines of doing string substitution in the `FileOrURI` trait using some form of recipe context, although making sure that calibration paths are always relative to `PYMEDATADIR` would be a potential short-term workaround. If we were to do re-writing (which I don't think we should), then scanning each module in the recipe for `FileOrURI` traits and globally re-writing those would be preferable to forcing recipe modules to register a hook. 
- Regardless of the final approach, I don't think we should make this subject to a config hook - we will just get mutually incompatible versions of the session files. 

All in all, it probably needs some further thought as to the best approach. To do this, I probably need a bit more context about what the general use cases for portable sessions are (and whether or not a lot of them could be solved by using, e.g. a network drive which is consistently mounted at the same place across computers).

David-Baddeley commented on PYME/LMVis/pipeline.py:
```diff
@@ -509,12 +509,13 @@ def _get_session_datasources(self):
 
             if hasattr(ds, 'filename'):
-                fn = ds.filename
+                fn = str(Path(ds.filename).resolve()) # ensure we return an absolute path!
```
This will mess with cluster URIs and paths relative to `PYMEDATADIR`

David-Baddeley commented:
Thinking a bit more, I wonder if this might be the expedient / pragmatic approach:

- don't modify (resolve) paths attached to datasources  (i.e. revert changes in `Pipeline._get_session_datasources`) to prevent issues with cluster URIs and `PYMEDATADIR` relative paths.
- In `VisCore.save_session()` check to see if **every** filename and every `FileOrURI` recipe trait is beneath the proposed session filename. If so, re-write as relative, and add a key to the session indicating that paths are relative.
- In `VisCore.load_session()` check for something like `if session.get('relative_paths', False):` and then rewrite all paths (including those in `FileOrURI` recipe traits) accordingly.

This should default to the current absolute (or non-portable) paths in all situations where it's not obviously clear that the session can be saved in a portable way. Maybe log whether you are saving a portable or non-portable session.

csoeller commented:
Thanks @David-Baddeley for your thoughts. I admit that cluster URIs and `PYMEDATADIR` had not featured in my thinking. I understand the general concern but just for my understanding, is there a realistic scenario where one would use sessions with cluster URIs and `PYMEDATADIR`? (even if there isn't an obvious one I am not arguing against the need to work with URIs etc to avoid surprises/hard to resolve user errors).

Note that in my code I wasn't changing the datasource filenames, just in `_get_session_datasources` I was resolving the paths that would be returned to build the session info.

I am therefore fully onboard with the proposed solution, we could even be one step more careful, if you think that could be useful, namely:

- don't modify (resolve) paths attached to datasources (i.e. revert changes in Pipeline._get_session_datasources) to prevent issues with cluster URIs and PYMEDATADIR relative paths.
- In VisCore.save_session() check to see if every filename and every FileOrURI recipe trait is beneath the proposed session filename. If so, re-write as relative, and add a key to the session indicating that paths are relative.
   - here we could instead write a session key with the full path to the session filename and leave all paths untouched as absolute
- In VisCore.load_session() check for something like if session.get('relative_paths', False): and then rewrite all paths (including those in FileOrURI recipe traits) accordingly.
   - instead, with paths left absolute in the written session but now a session key present that gives the full paths to the session file as it was written, we could do all checking upon loading, i.e.
      - if paths to datasources are not valid when trying to read, IF all existing paths are relative to (and below) the original file paths, then a final read attempt with all paths now translated relative to the *new* session file path could be attempted
      - if that also fails finally give up (with suitable error/warning message)

What do you think?

csoeller commented:
PS - the primary use case, and IMHO a valid one: the use of `.pvs` files and recipes with `FileOrURI` recipe traits is extermely annoying when we use OneDrive and DropBox shared folders for work on such files, as we do a lot in our lab.

In this everyday scenario in our use the absolute paths are all different (relative to local user directory) on different machines but the relative paths remain conserved across machines.

David-Baddeley commented:
I had initially thought of just saving the session path into the .pvs and working everything out on load failure (your suggested modification), but I don't think it's quite as clean ...

As I see it, the main advantage to doing the relative path translation on save is that you know if you have a portable .pvs before you try copy/moving and opening it. I also kind of like having the forced decision on relative or absolute paths rather than just falling back and trying to work out a relative path on failure as it makes it a bit clearer what the correct behaviour is. 

The one downside that I can think of is that we might need to insert some kind of token into the relative paths in the .pvs that we then replace on load i.e. saving as `$session_dir$/data_file.h5r` or `$session_dir$/some_calibration.json`. The upside of using the token approach is that it would make the loading super-easy - we'd  just need to do a `session_yaml.replace('$session_dir$', os.path.dirname(session_filename)` before we loaded it - we wouldn't strictly even need to check the `relative_paths` flag.

csoeller commented:
I like the session dir token approach quite a bit! Seems pretty clean to me and also human readable, if ever some manual editing were to occur.

csoeller committed:
Merge branch 'master' into session-paths-rewrite

csoeller committed:
first version that implements $session_dir$ token

- writes session dir token IF all ds paths below session path
- transition version that still keeps compatibility function to read previously written relative paths files without token
- does not touch URIs in any way
- reverts pipeline fiddling with ds filenames

csoeller committed:
fix typo

csoeller committed:
add SESSIONDIR_TOKEN constant to avoid typos

csoeller committed:
retrieve all FileOrURI parameter names from list of all modules

- this should enable recipe paths translation pretty automagically by recognising FileOrURI args

csoeller commented:
With latest commits the PR should now be close to the approach we had discussed:

- pipeline changes reverted
- use a sessiondir token as we had discussed
- only rewrite with session dir token if all data sources files are below session dir
- upon reading a session file automatically translate any present session token
   - note: currently it does not explicitly check for the `session['relative_paths']` flag, but we do record it correctly
- it runs through the list of all registered modules to get a list of all possible `FileOrURI` recipe parameters
   - incidentally a couple of modules could not be instantiated without raising a `TraitsError`: `failedmodules
['measurement.IntensityAtPoints', 'tracking.TrackFeatures']`

The PR should be ready for some initial comments on this round of commits.

**Caveats:**

- it still contains a compatibility function to load session files ok that do not yet contain the session dir token. This will allow me to rewrite the ones I already have sitting around. Will be removed when done...
- a bit more testing will surely be useful

csoeller committed:
add some debug logging; make compat loading config switchable for now

csoeller committed:
small comment tweaks

David-Baddeley commented:
I still don't like having modules have to register to get their paths changed - Is there a reason we need this?

David-Baddeley commented on PYME/LMVis/sessionpaths.py:
```diff
@@ -0,0 +1,161 @@
+    return fnamerel
+
+def allpaths_relative_to(session,sessiondir):
+    # currently we only check data sources, NOT recipe FileOrURI entries
```
We need to check FileOrURI too ....

csoeller replied:
Here it would be good to agree if one of the recipe FileOrURI arguments turning out to be a URI is reason enough to return False.

Am I incorrect in thinking that having a URI in a session to be saved is a bad idea in any case, paths relative or not?

David-Baddeley replied:
I think any paths that are not relative mean that the session is not portable (and it's much safer to be strict here).

I don't see any major issue in having URIs in sessions however (they are portable by default, assuming other settings are correct). In an ideal world, every path in the session would probably be a URI rather than an filesystem path. This assumes that all your computers are set up to use the same cluster(s), but that is the design intent.

David-Baddeley commented on PYME/LMVis/visCore.py:
```diff
@@ -847,8 +870,15 @@ def OnSaveSession(self, event):
+        sessiondir = fpath.resolve().parent # note that filename could be relative, in that case need to resolve to obtain session dir
+        session = yaml.safe_load(fpath.read_text().replace(SESSIONDIR_TOKEN,str(sessiondir))) # replace any possibly present SESSIONDIR_TOKEN
+
+        # temporary compatibility call
```
Need to remove any compat stuff before merge

csoeller commented:
> I still don't like having modules have to register to get their paths changed - Is there a reason we need this?

No, this can go with the latest changes. It is not used anymore.

David-Baddeley commented:
As it stands, you still walk all the loaded modules and register modules with `FileOrURI` traits - I think this would be cleaner if detection were done on the fly -  potentially assisted by a classmethod on the base recipe module class e.g.

```python
@classmethod
    def file_or_uri_traits(cls):
        """Return a list of FileOrURI traits, as these can require special handling (potentially warn if this list is not empty)"""
        return [k for k, v in cls.class_traits.items() if isinstance(v.trait_type, traits.FileOrURI)]
``` 

and in the session code:
```python
file_inputs = []
for mn, mod_dict in recipe:
    file_traits = base.all_modules[mn].file_or_uri_traits()
    file_inputs.extend([mod_dict[k] for k in file_traits]
```

csoeller committed:
remove compat stuff; add recipe check and implement classmethod

csoeller commented:
Al the main things should now be in as mentioned above; could possibly coded more cleverly in some of the loops but appears to do the job.

csoeller committed:
some tyding

csoeller committed:
a few more checks

csoeller commented:
@David-Baddeley , all features we discussed should now be implemented. I have also run some tests with a notebook and some session manipulations. For the cases tested it seems ok. It should be good for another check if it is ready for merging now.

David-Baddeley commented on PYME/LMVis/sessionpaths.py:
```diff
@@ -0,0 +1,86 @@
+            return False
+    for module in session['recipe']:
+        [(mn, mod_dict)] = module.items()
+        # we only check registered modules - session loading will anyway fail with an unregistered module
```
Is it really possible that there is an un-registered module in the recipe??? If that is the case, is it not better that the session saving fails?

David-Baddeley committed:
some tidying and documentation

csoeller commented:
Many thanks @David-Baddeley for review, further tweaks and merging. One issue is to be fixed with your merged PR, details in #1548. Would be cool if that could be merged (with suitable tweaks if required) so that it then works without raising an error.
